File: rag_poc_novatech.ts
import * as fs from "fs/promises";
import { existsSync } from "fs";
import * as path from "path";
import * as net from "net";

import { ChromaClient } from "chromadb";
import { pipeline } from "@xenova/transformers";

type Metadata = {
  document: string;
  chunk_index: number;
};

type QueryResult = {
  chunkId: string;
  document: string;
  text: string;
  score: number;
};

type InMemoryChunk = {
  id: string;
  text: string;
  metadata: Metadata;
  embedding: number[];
};

const __dirname = process.cwd();

const COLLECTION_NAME = "novatech_rag";
const MAX_WORDS = 320;
const OVERLAP_WORDS = 40;
const EMBEDDING_MODEL = "Xenova/all-MiniLM-L6-v2";

const FALLBACK_DOCUMENTS: Array<{ name: string; content: string }> = [
  {
    name: "POL-001-politica-devolucao.md",
    content:
      "POL-001 seção 3.2: mercadorias podem ser devolvidas em até 7 dias úteis após recebimento, exceto cargas perigosas (classes 1 a 6 ANTT).",
  },
  {
    name: "SLA-2024-tabela-sla-clientes.md",
    content:
      "SLA-2024: cliente Gold resposta em até 2h e resolução em até 24h; cliente Silver resposta em até 4h e resolução em até 48h; cliente Standard resposta em até 8h e resolução em até 72h.",
  },
  {
    name: "PROC-042-v2-frete-especial-revisado.md",
    content:
      "PROC-042-v2 seção 2: frete especial acima de 500kg = valor base x multiplicador regional. Norte 1.8, Sudeste 1.1, Sul 1.3, Nordeste 1.5, Centro-Oeste 1.4.",
  },
];

const inMemoryStore: InMemoryChunk[] = [];

async function readDocument(filePath: string): Promise<string> {
  return fs.readFile(filePath, "utf-8");
}

function simpleChunkText(text: string, maxWords = MAX_WORDS, overlapWords = OVERLAP_WORDS): string[] {
  const words = text.split(/\s+/).filter(Boolean);
  if (words.length === 0) {
    return [];
  }

  const chunks: string[] = [];
  let start = 0;

  while (start < words.length) {
    const end = Math.min(words.length, start + maxWords);
    const chunk = words.slice(start, end).join(" ").trim();
    if (chunk.length > 0) {
      chunks.push(chunk);
    }

    if (end >= words.length) {
      break;
    }

    start = Math.max(end - overlapWords, start + 1);
  }

  return chunks;
}

let embeddingPipeline: any = null;

async function getEmbeddingPipeline() {
  if (!embeddingPipeline) {
    embeddingPipeline = await pipeline("feature-extraction", EMBEDDING_MODEL);
  }
  return embeddingPipeline;
}

async function embedTexts(texts: string[]): Promise<number[][]> {
  const extractor = await getEmbeddingPipeline();
  const vectors: number[][] = [];

  for (const text of texts) {
    const output: any = await extractor(text, { pooling: "mean", normalize: true } as any);
    vectors.push(Array.from(output.data as Float32Array));
  }

  return vectors;
}

async function getCollection() {
  const client: any = new ChromaClient({
    host: "localhost",
    port: 8000,
    ssl: false,
  });

  try {
    return await client.getCollection({ name: COLLECTION_NAME } as any);
  } catch {
    return await client.createCollection({ name: COLLECTION_NAME } as any);
  }
}

async function isChromaReachable(host = "127.0.0.1", port = 8000): Promise<boolean> {
  return new Promise((resolve) => {
    const socket = new net.Socket();

    socket.setTimeout(500);

    socket.once("connect", () => {
      socket.destroy();
      resolve(true);
    });

    socket.once("error", () => {
      socket.destroy();
      resolve(false);
    });

    socket.once("timeout", () => {
      socket.destroy();
      resolve(false);
    });

    socket.connect(port, host);
  });
}

function cosineSimilarity(a: number[], b: number[]): number {
  let dot = 0;
  let normA = 0;
  let normB = 0;

  for (let i = 0; i < a.length; i++) {
    dot += a[i] * b[i];
    normA += a[i] * a[i];
    normB += b[i] * b[i];
  }

  if (normA === 0 || normB === 0) {
    return 0;
  }

  return dot / (Math.sqrt(normA) * Math.sqrt(normB));
}

async function ingestDocuments(documentPaths: string[]): Promise<void> {
  const ids: string[] = [];
  const documents: string[] = [];
  const metadatas: Metadata[] = [];

  for (const docPath of documentPaths) {
    const rawText = await readDocument(docPath);
    const chunks = simpleChunkText(rawText);

    chunks.forEach((chunk, index) => {
      ids.push(`${path.parse(docPath).name}-${index}`);
      documents.push(chunk);
      metadatas.push({
        document: path.basename(docPath),
        chunk_index: index,
      });
    });
  }

  if (documents.length === 0) {
    throw new Error("Nenhum documento válido foi encontrado para ingestão.");
  }

  const embeddings = await embedTexts(documents);

  let collection: any = null;
  const chromaOnline = await isChromaReachable();
  if (chromaOnline) {
    collection = await getCollection();
  }

  if (!collection) {
    // Fallback para execução local sem servidor Chroma.
    for (let i = 0; i < ids.length; i++) {
      inMemoryStore.push({
        id: ids[i],
        text: documents[i],
        metadata: metadatas[i],
        embedding: embeddings[i],
      });
    }
    return;
  }

  await collection.add({
    ids,
    documents,
    metadatas,
    embeddings,
  });
}

async function search(question: string, topK = 5): Promise<QueryResult[]> {
  const [queryEmbedding] = await embedTexts([question]);

  let collection: any = null;
  const chromaOnline = await isChromaReachable();
  if (chromaOnline) {
    collection = await getCollection();
  }

  if (!collection) {
    const ranked = inMemoryStore
      .map((item) => ({
        item,
        score: cosineSimilarity(queryEmbedding, item.embedding),
      }))
      .sort((a, b) => b.score - a.score)
      .slice(0, topK);

    return ranked.map(({ item, score }) => ({
      chunkId: `${item.metadata.document}#${item.metadata.chunk_index}`,
      document: item.metadata.document,
      text: item.text,
      score,
    }));
  }

  const result: any = await collection.query({
    queryEmbeddings: [queryEmbedding],
    nResults: topK,
  });

  const docs = (result.documents?.[0] ?? []) as Array<string | null>;
  const metas = (result.metadatas[0] ?? []) as Metadata[];
  const distances = (result.distances?.[0] ?? []) as number[];

  return docs.map((doc: string | null, index: number) => {
    const metadata = metas[index];
    const distance = distances[index] ?? 1;
    return {
      chunkId: `${metadata.document}#${metadata.chunk_index}`,
      document: metadata.document,
      text: doc ?? "",
      score: 1 - Number(distance),
    };
  });
}

function buildPrompt(question: string, chunks: QueryResult[]): string {
  const contextBlock = chunks.map((chunk) => `[${chunk.chunkId}] ${chunk.text}`).join("\n\n");

  return `Você é o assistente de atendimento da NovaTech.

Regras:
- Use apenas o contexto fornecido.
- Cite a fonte.
- Não invente valores, prazos ou regras.
- Se a informação não estiver no contexto, diga que não encontrou.
- Se houver conflito entre versões, explicite a contradição.

Contexto recuperado:
${contextBlock}

Pergunta:
${question}

Resposta:`;
}

function loadDefaultDocuments(): string[] {
  const files = [
    "POL-001-politica-devolucao.md",
    "PROC-042-frete-especial-v1.md",
    "PROC-042-v2-frete-especial-revisado.md",
    "SLA-2024-tabela-sla-clientes.md",
    "FAQ-atendimento.md",
  ];

  return files.map((name) => {
    const localPath = path.join(__dirname, name);
    if (existsSync(localPath)) {
      return localPath;
    }

    return path.join(__dirname, "..", name);
  });
}

async function ingestFallbackDocuments(): Promise<void> {
  const ids: string[] = [];
  const docs: string[] = [];
  const metadatas: Metadata[] = [];

  FALLBACK_DOCUMENTS.forEach((doc, index) => {
    ids.push(`${path.parse(doc.name).name}-${index}`);
    docs.push(doc.content);
    metadatas.push({
      document: doc.name,
      chunk_index: index,
    });
  });

  const embeddings = await embedTexts(docs);
  for (let i = 0; i < ids.length; i++) {
    inMemoryStore.push({
      id: ids[i],
      text: docs[i],
      metadata: metadatas[i],
      embedding: embeddings[i],
    });
  }
}

async function main() {
  const docs = loadDefaultDocuments().filter((p) => existsSync(p));

  if (docs.length > 0) {
    await ingestDocuments(docs);
  } else {
    await ingestFallbackDocuments();
  }

  const question = "Qual o prazo de devolução para carga perigosa?";
  const chunks = await search(question, 3);
  const prompt = buildPrompt(question, chunks);

  console.log(prompt);
}

main().catch((error) => {
  console.error("Erro ao executar a POC de RAG:", error);
  process.exit(1);
});

File: semana-02/3-resultados/projeto/rag_poc_novatech.py
import os
import socket
import math
from pathlib import Path
from typing import TypedDict

from sentence_transformers import SentenceTransformer
import chromadb

# ---------------------------------------------------------------------------
# Configurações
# ---------------------------------------------------------------------------

COLLECTION_NAME = "novatech_rag"
MAX_WORDS = 320
OVERLAP_WORDS = 40
EMBEDDING_MODEL = "all-MiniLM-L6-v2"

FALLBACK_DOCUMENTS = [
    {
        "name": "POL-001-politica-devolucao.md",
        "content": (
            "POL-001 seção 3.2: mercadorias podem ser devolvidas em até 7 dias úteis "
            "após recebimento, exceto cargas perigosas (classes 1 a 6 ANTT)."
        ),
    },
    {
        "name": "SLA-2024-tabela-sla-clientes.md",
        "content": (
            "SLA-2024: cliente Gold resposta em até 2h e resolução em até 24h; "
            "cliente Silver resposta em até 4h e resolução em até 48h; "
            "cliente Standard resposta em até 8h e resolução em até 72h."
        ),
    },
    {
        "name": "PROC-042-v2-frete-especial-revisado.md",
        "content": (
            "PROC-042-v2 seção 2: frete especial acima de 500kg = valor base x multiplicador regional. "
            "Norte 1.8, Sudeste 1.1, Sul 1.3, Nordeste 1.5, Centro-Oeste 1.4."
        ),
    },
]

# ---------------------------------------------------------------------------
# Tipos
# ---------------------------------------------------------------------------

class Metadata(TypedDict):
    document: str
    chunk_index: int


class QueryResult(TypedDict):
    chunk_id: str
    document: str
    text: str
    score: float


class InMemoryChunk(TypedDict):
    id: str
    text: str
    metadata: Metadata
    embedding: list[float]


# ---------------------------------------------------------------------------
# Store em memória (fallback quando ChromaDB não está disponível)
# ---------------------------------------------------------------------------

in_memory_store: list[InMemoryChunk] = []

# ---------------------------------------------------------------------------
# Modelo de embeddings (singleton)
# ---------------------------------------------------------------------------

_embedding_model: SentenceTransformer | None = None


def get_embedding_model() -> SentenceTransformer:
    global _embedding_model
    if _embedding_model is None:
        _embedding_model = SentenceTransformer(EMBEDDING_MODEL)
    return _embedding_model


def embed_texts(texts: list[str]) -> list[list[float]]:
    model = get_embedding_model()
    vectors = model.encode(texts, normalize_embeddings=True)
    return [v.tolist() for v in vectors]

# ---------------------------------------------------------------------------
# Chunking
# ---------------------------------------------------------------------------

def simple_chunk_text(text: str, max_words: int = MAX_WORDS, overlap_words: int = OVERLAP_WORDS) -> list[str]:
    words = text.split()
    if not words:
        return []

    chunks: list[str] = []
    start = 0

    while start < len(words):
        end = min(len(words), start + max_words)
        chunk = " ".join(words[start:end]).strip()
        if chunk:
            chunks.append(chunk)
        if end >= len(words):
            break
        start = max(end - overlap_words, start + 1)

    return chunks

# ---------------------------------------------------------------------------
# ChromaDB
# ---------------------------------------------------------------------------

def is_chroma_reachable(host: str = "127.0.0.1", port: int = 8000) -> bool:
    try:
        with socket.create_connection((host, port), timeout=0.5):
            return True
    except OSError:
        return False


def get_collection():
    client = chromadb.HttpClient(host="localhost", port=8000)
    try:
        return client.get_collection(name=COLLECTION_NAME)
    except Exception:
        return client.create_collection(name=COLLECTION_NAME)

# ---------------------------------------------------------------------------
# Similaridade cosseno (para o fallback in-memory)
# ---------------------------------------------------------------------------

def cosine_similarity(a: list[float], b: list[float]) -> float:
    dot = sum(x * y for x, y in zip(a, b))
    norm_a = math.sqrt(sum(x * x for x in a))
    norm_b = math.sqrt(sum(x * x for x in b))
    if norm_a == 0 or norm_b == 0:
        return 0.0
    return dot / (norm_a * norm_b)

# ---------------------------------------------------------------------------
# Ingestão
# ---------------------------------------------------------------------------

def ingest_documents(document_paths: list[str]) -> None:
    ids: list[str] = []
    documents: list[str] = []
    metadatas: list[Metadata] = []

    for doc_path in document_paths:
        raw_text = Path(doc_path).read_text(encoding="utf-8")
        chunks = simple_chunk_text(raw_text)
        stem = Path(doc_path).stem

        for index, chunk in enumerate(chunks):
            ids.append(f"{stem}-{index}")
            documents.append(chunk)
            metadatas.append({"document": Path(doc_path).name, "chunk_index": index})

    if not documents:
        raise ValueError("Nenhum documento válido foi encontrado para ingestão.")

    embeddings = embed_texts(documents)

    if is_chroma_reachable():
        collection = get_collection()
        collection.add(ids=ids, documents=documents, metadatas=metadatas, embeddings=embeddings)
    else:
        for i in range(len(ids)):
            in_memory_store.append({
                "id": ids[i],
                "text": documents[i],
                "metadata": metadatas[i],
                "embedding": embeddings[i],
            })


def ingest_fallback_documents() -> None:
    ids: list[str] = []
    docs: list[str] = []
    metadatas: list[Metadata] = []

    for index, doc in enumerate(FALLBACK_DOCUMENTS):
        stem = Path(doc["name"]).stem
        ids.append(f"{stem}-{index}")
        docs.append(doc["content"])
        metadatas.append({"document": doc["name"], "chunk_index": index})

    embeddings = embed_texts(docs)
    for i in range(len(ids)):
        in_memory_store.append({
            "id": ids[i],
            "text": docs[i],
            "metadata": metadatas[i],
            "embedding": embeddings[i],
        })

# ---------------------------------------------------------------------------
# Busca
# ---------------------------------------------------------------------------

def search(question: str, top_k: int = 5) -> list[QueryResult]:
    query_embedding = embed_texts([question])[0]

    if is_chroma_reachable():
        collection = get_collection()
        result = collection.query(query_embeddings=[query_embedding], n_results=top_k)
        docs = result["documents"][0] if result.get("documents") else []
        metas = result["metadatas"][0] if result.get("metadatas") else []
        distances = result["distances"][0] if result.get("distances") else []

        results: list[QueryResult] = []
        for i, doc in enumerate(docs):
            meta: Metadata = metas[i]
            distance = distances[i] if i < len(distances) else 1.0
            results.append({
                "chunk_id": f"{meta['document']}#{meta['chunk_index']}",
                "document": meta["document"],
                "text": doc or "",
                "score": 1 - float(distance),
            })
        return results

    # Fallback in-memory
    ranked = sorted(
        in_memory_store,
        key=lambda item: cosine_similarity(query_embedding, item["embedding"]),
        reverse=True,
    )[:top_k]

    return [
        {
            "chunk_id": f"{item['metadata']['document']}#{item['metadata']['chunk_index']}",
            "document": item["metadata"]["document"],
            "text": item["text"],
            "score": cosine_similarity(query_embedding, item["embedding"]),
        }
        for item in ranked
    ]

# ---------------------------------------------------------------------------
# Prompt
# ---------------------------------------------------------------------------

def build_prompt(question: str, chunks: list[QueryResult]) -> str:
    context_block = "\n\n".join(f"[{c['chunk_id']}] {c['text']}" for c in chunks)

    return f"""Você é o assistente de atendimento da NovaTech.

Regras:
- Use apenas o contexto fornecido.
- Cite a fonte.
- Não invente valores, prazos ou regras.
- Se a informação não estiver no contexto, diga que não encontrou.
- Se houver conflito entre versões, explicite a contradição.

Contexto recuperado:
{context_block}

Pergunta:
{question}

Resposta:"""

# ---------------------------------------------------------------------------
# Documentos padrão
# ---------------------------------------------------------------------------

def load_default_documents() -> list[str]:
    files = [
        "POL-001-politica-devolucao.md",
        "PROC-042-frete-especial-v1.md",
        "PROC-042-v2-frete-especial-revisado.md",
        "SLA-2024-tabela-sla-clientes.md",
        "FAQ-atendimento.md",
    ]

    base_dir = Path(__file__).parent
    paths: list[str] = []

    for name in files:
        local = base_dir / name
        parent = base_dir.parent / name
        if local.exists():
            paths.append(str(local))
        elif parent.exists():
            paths.append(str(parent))

    return paths

# ---------------------------------------------------------------------------
# Main
# ---------------------------------------------------------------------------

def main() -> None:
    docs = load_default_documents()

    if docs:
        ingest_documents(docs)
    else:
        ingest_fallback_documents()

    question = "Qual o prazo de devolução para carga perigosa?"
    chunks = search(question, top_k=3)
    prompt = build_prompt(question, chunks)

    print(prompt)


if __name__ == "__main__":
    main()
